### PR TITLE
Company Emotion Functions

### PR DESCRIPTION
**Pull request by duncanspumpkin**
Continue from #2286

Cleans up how company emotions work and implements the update emotion function.

### CODE REVIEW

File: src/OpenLoco/src/World/Company.cpp [hunk 1]
```diff
@@ -237,6 +261,62 @@ namespace OpenLoco
...
+            company->observationEntity = entity;
+            company->observationObject = object;
+            company->observationStatus = status;
+            auto closestTown = TownManager::getClosestTownAndDensity(pos);
+            if (closestTown.has_value())
+            {
+                company->observationTownId = closestTown->first;
+            }
+            Ui::WindowManager::invalidate(Ui::WindowType::company, enumValue(id));
+            Ui::WindowManager::invalidate(Ui::WindowType::companyList, enumValue(id));
```
Comment (AaronVanGeffen) [inline src/OpenLoco/src/World/Company.cpp:315]: Is the company list window ever bound to a company id?

File: src/OpenLoco/src/World/CompanyAi.cpp [hunk 2]
```diff
@@ -968,4 +970,62 @@ namespace OpenLoco
...
+            auto& thought = company->aiThoughts[company->var_2578];
+            if (kThoughtTypeFlags[enumValue(thought.type)] & (1U << 1))
+            {
+                auto* industry = IndustryManager::get(static_cast<IndustryId>(thought.var_01));
+                pos = World::Pos2{ industry->x, industry->y };
+            }
+            else
+            {
+                // Interestingly var_01 isn't a uint16_t
+                auto* town = TownManager::get(static_cast<TownId>(thought.var_01));
```
Comment (AaronVanGeffen) [inline src/OpenLoco/src/World/CompanyAi.cpp:990]: Doesn't really matter though. From S5/Limits.h:
```
    constexpr size_t kMaxTowns = 80;
```

File: src/OpenLoco/src/World/Company.h [hunk 3]
```diff
@@ -193,9 +225,11 @@ namespace OpenLoco
...
     static_assert(offsetof(Company, vehicleProfit) == 0x8B9E);
     static_assert(offsetof(Company, challengeProgress) == 0x8C4E);
-    static_assert(offsetof(Company, var_8BB0) == 0x8BB0);
+    static_assert(offsetof(Company, activeEmotions) == 0x8BB0);
 
     StringId getCorporateRatingAsStringId(CorporateRating rating);
     constexpr CorporateRating performanceToRating(int16_t performanceIndex);
     void formatPerformanceIndex(const int16_t performanceIndex, FormatArguments& args);
+    void companyEmotionEvent(CompanyId companyId, Emotion emotion);
+    void companySetObservation(CompanyId id, ObservationStatus status, World::Pos2 pos, EntityId entity, uint16_t object);
```
[the remark this answers is not in the document]
Comment (duncanspumpkin) [reply src/OpenLoco/src/World/Company.h:234]: In general i've been trying to reduce the numbers of member functions we have. I much more lean towards having POD types and functions.

**Review by AaronVanGeffen — APPROVED** [review]
Appears to be working well. Just to be on the safe side, have you performed the usual replay tests?

**Comment by duncanspumpkin** [comment]
> Appears to be working well. Just to be on the safe side, have you performed the usual replay tests?

... I have now.

**Pull request merged**